store/sqlstore/post_store.go

```diff
@@ -72,42 +99,85 @@ func (s *SqlPostStore) CreateIndexesIfNotExists() {
+		}
+		post.PreSave()
+		maxPostSize := s.GetMaxPostSize()
+		if err := post.IsValid(maxPostSize); err != nil {
+			return nil, err
+		}
 
-	maxPostSize := s.GetMaxPostSize()
+		currentChannelCount, ok := channelNewPosts[post.ChannelId]
+		if !ok {
+			if post.IsJoinLeaveMessage() {
+				channelNewPosts[post.ChannelId] = 0
+			} else {
+				channelNewPosts[post.ChannelId] = 1
+			}
+			maxDateNewPosts[post.ChannelId] = post.CreateAt
+		} else {
+			if !post.IsJoinLeaveMessage() {
+				channelNewPosts[post.ChannelId] = currentChannelCount + 1
+			}
+			if post.CreateAt > maxDateNewPosts[post.ChannelId] {
+				maxDateNewPosts[post.ChannelId] = post.CreateAt
+			}
+		}
 
-	post.PreSave()
-	if err := post.IsValid(maxPostSize); err != nil {
-		return nil, err
+		if len(post.RootId) > 0 {
+			currentRootCount, ok := rootIds[post.RootId]
+			if !ok {
+				rootIds[post.RootId] = 1
+				maxDateRootIds[post.RootId] = post.CreateAt
+			} else {
+				rootIds[post.RootId] = currentRootCount + 1
+				if post.CreateAt > maxDateRootIds[post.RootId] {
+					maxDateRootIds[post.RootId] = post.CreateAt
+				}
+			}
+		}
 	}
 
-	if err := s.GetMaster().Insert(post); err != nil {
-		return nil, model.NewAppError("SqlPostStore.Save", "store.sql_post.save.app_error", nil, "id="+post.Id+", "+err.Error(), http.StatusInternalServerError)
+	query := s.getQueryBuilder().Insert("Posts").Columns(postSliceColumns()...)
+	for _, post := range posts {
+		query = query.Values(postToSlice(post)...)
+	}
+	sql, args, err := query.ToSql()
+	if err != nil {
+		return nil, model.NewAppError("SqlPostStore.Save", "store.sql_post.save.app_error", nil, err.Error(), http.StatusInternalServerError)
 	}
 
-	time := post.UpdateAt
+	if _, err := s.GetMaster().Exec(sql, args...); err != nil {
+		return nil, model.NewAppError("SqlPostStore.Save", "store.sql_post.save.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
 
-	if !post.IsJoinLeaveMessage() {
-		if _, err := s.GetMaster().Exec("UPDATE Channels SET LastPostAt = GREATEST(:LastPostAt, LastPostAt), TotalMsgCount = TotalMsgCount + 1 WHERE Id = :ChannelId", map[string]interface{}{"LastPostAt": time, "ChannelId": post.ChannelId}); err != nil {
-			mlog.Error("Error updating Channel LastPostAt.", mlog.Err(err))
-		}
-	} else {
-		// don't update TotalMsgCount for unimportant messages so that the channel isn't marked as unread
-		if _, err := s.GetMaster().Exec("UPDATE Channels SET LastPostAt = :LastPostAt WHERE Id = :ChannelId AND LastPostAt < :LastPostAt", map[string]interface{}{"LastPostAt": time, "ChannelId": post.ChannelId}); err != nil {
+	for channelId, count := range channelNewPosts {
+		if _, err := s.GetMaster().Exec("UPDATE Channels SET LastPostAt = GREATEST(:LastPostAt, LastPostAt), TotalMsgCount = TotalMsgCount + :Count WHERE Id = :ChannelId", map[string]interface{}{"LastPostAt": maxDateNewPosts[channelId], "ChannelId": channelId, "Count": count}); err != nil {
```

Comment:
It is possible to do this in a single query. See: ChannelStore.UpdateLastViewedAt

Reply:
Yes I think so, I'll take a look.

Reply:
I don't think is an amazing approach, because the current idea is "fire and forget" and I prefer to have 1 small transaction failing than all of them failing as a whole and logging the error.

Maybe we can add this updates inside a gorouting to avoid waiting until is completed, but I don't see much value there, and the performance improvements around that are minimal.

Reply:
The problem here is we have to update different channels with different dates, so we can follow other approach to get the LastPostAt using something like `UPDATE Channels SET LastPostAt = (CASE WHEN ChannelId = X THEN Y WHEN ChannelId = Z THEN J) WHERE ChannelId IN (X, Z)`. But I don't see the point, here we are trying to maintain the consistency, and this error shouldn't happend, and probably is going to be autofixed quickly (with the next post in the channel).

Reply:
There are 2 separate concerns which I had:

> But I don't see the point, here we are trying to maintain the consistency, and this error shouldn't happend,

1. The point, IMO is the same reason why we are doing `SaveMultiple` - to prevent multiple roundtrips from the database. I am pretty confident that running 1 query would be faster than running 10 individual queries. Otherwise we could just run `Save` in a loop in a transaction. This is unrelated to the idempotency issue.

2. Idempotency. Your point about it going to be autofixed is totally valid, but I was talking about the posts table. Let's say channel update fails, and the user tries the whole bulk import again, wouldn't it create duplicate posts in the DB, because SaveMultiple would run again ?

Reply:
Filed https://mattermost.atlassian.net/browse/MM-22974 for 1. 2 was discussed offline and it's fine.